lib/experimental/html/class-wp-html-character-reference-transcoder.php

```diff
@@ -0,0 +1,1149 @@
+					( 0xFFFE === ( $code_point & 0xFFFE ) ) ||
+
+					// 0x0D or non-ASCII-whitespace control
+					0x0D === $code_point ||
+					(
+						$code_point >= 0 &&
+						$code_point <= 0x1F &&
+						0x9 !== $code_point &&
+						0xA !== $code_point &&
+						0xC !== $code_point &&
+						0xD !== $code_point
+					)
+				) {
+					$at += $digit_count;
+					$buffer .= substr( $input, $next, $at - $next );
+					continue;
+				}
+
+				if ( $code_point >= 0x80 && $code_point <= 0x9F && array_key_exists( $code_point, self::$c1_replacements ) ) {
+					$at += $digit_count;
+					if ( $at < $end && ';' === $input[ $at ] ) {
+						$at++;
+					}
+					$buffer .= self::$c1_replacements[ $code_point ];
+					continue;
+				}
+
+				// Convert code point to UTF-8 bytes.
+				if ( $code_point < 0x80 ) {
+					$buffer .= sprintf( '%c', $code_point & 0x7F );
+				} else if ( $code_point < 0x800 ) {
+					$buffer .= sprintf(
+						'%c%c',
+						0xC0 | ( ( $code_point >> 6 ) & 0x1F ),
+						0x80 | ( $code_point & 0x3F )
+					);
+				} else if ( $code_point < 0x10000 ) {
+					$buffer .= sprintf(
+						'%c%c%c',
+						0xE0 | ( ( $code_point >> 12 ) & 0x0F ),
+						0x80 | ( ( $code_point >> 6 ) & 0x3F ),
+						0x80 | ( $code_point & 0x3F )
+					);
+				} else {
+					$buffer .= sprintf(
+						'%c%c%c%c',
+						0xF0 | ( ( $code_point >> 18 ) & 0x07 ),
+						0x80 | ( ( $code_point >> 12 ) & 0x3F ),
+						0x80 | ( ( $code_point >> 6 ) & 0x3F ),
+						0x80 | ( $code_point & 0x3F )
+					);
+				}
+
+				$at += $digit_count;
+				if ( $at < $end && ';' === $input[ $at ] ) {
+					$at++;
+				}
+				continue;
+			}
+
+			// &Aacute; -> group "Aa" (skip & since we know it's there).
+			$group = substr( $input, $at, 2 );
+
+			if ( array_key_exists( $group, self::$lookup_table ) ) {
+				$at += 2;
+
+				$group = self::$lookup_table[ $group ];
+
+				$i = 0;
+				while ( $i < strlen( $group ) ) {
+					/*
+					 * Extract name and substitution information from group string.
+					 *
+					 * Example:
+					 *
+					 * For group "qu", during lookup that will find "&quot;"
+					 *
+					 * ┌─────┬────┬──────┬────┬──────────────┬────┬─────┐
+					 * │ ... │ N5 │ Name │ S5 │ Substitution │ N6 │ ... │
+					 * ├─────┼────┼──────┼────┼──────────────┼────┼─────┤
+					 * │ ... │ 04 │ ote; │ 01 │ "            │ 03 │ ... │
+					 * └─────┴────┴──────┴────┴──────────────┴────┴─────┘
+					 *         ^^          ^^
+					 *          |           |
+					 *          |           ╰ The substitution is one byte,
+					 *          |             even though it's represented in
+					 *          |             the string literal as "\x22", which
+					 *          |             is done for the sake of avoiding
+					 *          |             quoting issues in PHP.
+					 *          |
+					 *          ╰ The "ote;" is four bytes (the finishing of &quo̱ṯe̱;̱)
+					 *
+					 * The part of the group string this represents follows:
+					 * > ...\x04ote;\x01\x22\x03...
```

Comment:
I'm confused with this comment. It says the group `qu` will find `&quot;`, but the example talks about `&quote;`. What does `N5`, `S5`, and `N6` mean in the header? I figured out the group string contains different possible substitutions to reduce the size of the lookup table, but I'm not sure what else is there to know about it.

Reply:
I looked up that entry in `$lookup_table` and it differs from what's described in this comment:

```
"QU" => "\x03OT;\x01\x22\x02OT\x01\x22",           // &QUOT; &QUOT
```

Reply:
the `ote;` is a typo, but you're not looking up the listed entry, since you looked up `QU` instead of `qu`. `QU` might be a better example though since it only has the two entries and this one has more.

```
// &quaternions; &quatint; &questeq; &quest; &quot; &quot
"qu" => "\x0aaternions;\x03ℍ\x06atint;\x03⨖\x06esteq;\x03≟\x04est;\x01?\x03ot;\x01\x22\x02ot\x01\x22",
```

still working on the table; trying to show the structural layout of the lookups